service/controller.go

```diff
@@ -710,54 +749,66 @@ func (s *service) createVolumeFromSnapshot(ctx context.Context, isiConfig *Isilo
+	if size > sizeInBytes {
+		return fmt.Errorf("specified size '%d' is smaller than source volume size '%d'", sizeInBytes, size)
+	}
+
+	srcPath := path.Join(isiPath, srcVolumeName)
+	snapshotName := utils.GetInternalSnapshotName(srcVolumeName, dstVolumeName)
+
+	var snapshot isi.Snapshot
+	if snapshot, err = isiConfig.isiSvc.CreateSnapshot(ctx, srcPath, snapshotName); err != nil {
+		return fmt.Errorf("failed to create intermediate snapshot of '%s' with name '%s', error: '%v'", srcPath, snapshotName, err)
+	}
+
+	_, err = isiConfig.isiSvc.CreateWritableSnapshot(ctx, strconv.FormatInt(snapshot.ID, 10), isiPath, dstVolumeName)
+	if err != nil {
+		isiConfig.isiSvc.DeleteSnapshot(ctx, snapshot.ID, snapshotName)
+		return fmt.Errorf("failed to create writable snapshot of '%s' with name '%s', error: '%v'", srcPath, snapshotName, err)
 	}
 
 	return nil
 }
 
-func (s *service) createVolumeFromSource(
+func (s *service) copyVolumeFromSource(
 	ctx context.Context,
 	isiConfig *IsilonClusterConfig,
 	isiPath string,
 	contentSource *csi.VolumeContentSource,
 	req *csi.CreateVolumeRequest,
 	sizeInBytes int64, accessZone string,
 ) error {
-	if contentSnapshot := contentSource.GetSnapshot(); contentSnapshot != nil {
-		// create volume from source snapshot
-		if err := s.createVolumeFromSnapshot(ctx, isiConfig, isiPath, contentSnapshot.GetSnapshotId(), req.GetName(), sizeInBytes, accessZone); err != nil {
-			return status.Error(codes.Internal, err.Error())
-		}
+	sourceVolume := contentSource.GetVolume()
+	if sourceVolume == nil {
+		return fmt.Errorf("source volume is nil")
 	}
 
-	if contentVolume := contentSource.GetVolume(); contentVolume != nil {
-		// create volume from source volume
-		srcVolumeName, _, _, _, err := utils.ParseNormalizedVolumeID(ctx, contentVolume.GetVolumeId())
-		if err != nil {
-			return status.Error(codes.Internal, err.Error())
-		}
-		if err := s.createVolumeFromVolume(ctx, isiConfig, isiPath, srcVolumeName, req.GetName(), sizeInBytes); err != nil {
-			return status.Error(codes.Internal, err.Error())
-		}
+	srcVolumeName, _, _, _, err := utils.ParseNormalizedVolumeID(ctx, sourceVolume.GetVolumeId())
+	if err != nil {
+		return status.Error(codes.Internal, err.Error())
+	}
+
+	if err := s.cloneVolume(ctx, isiConfig, isiPath, srcVolumeName, req.GetName(), sizeInBytes, accessZone); err != nil {
+		return status.Error(codes.Internal, err.Error())
 	}
+
 	return nil
 }
 
-func (s *service) getCreateVolumeResponse(ctx context.Context, exportID int, volName, path, accessZone string, sizeInBytes int64, azServiceIP, rootClientEnabled, sourceSnapshotID, sourceVolumeID, clusterName string) *csi.CreateVolumeResponse {
+// newCreateVolumeResponse creates a new CreateVolumeResponse instance.
+func (s *service) newCreateVolumeResponse(ctx context.Context, exportID int, volName, path, accessZone string, sizeInBytes int64, azServiceIP, rootClientEnabled, sourceSnapshotID, sourceVolumeID, clusterName string) *csi.CreateVolumeResponse {
```

Comment:
good to have a volume attribute to tell if this is a writable snapshot

Reply:
Added an attribute, VolumeType. Values: Directory, Snapshot, WritableSnapshot.